DINT-317: custom events and fields (#136)
adobe/magento-storefront-event-collector: README.md

## Overview

> This package listens for and handles events published by the [Magento Storefront Events SDK][sdk] to the adobeDataLayer[acdl]. It runs as a side effect and is meant to be a convenience for users who want to send tracking events to Adobe Commerce.

Reviewer: In product recommendations, I was discouraged from using terms like `tracking` and instead use `collecting`. But I wonder in this context, if we could reword to: "It runs as a side effect and is meant to be a convenience for users who want to send event data to Adobe Commerce."

    `StorefrontInstance`

    Events not forwarded to Adobe Commerce:

    `createAccount`

    `editAccount`

    `signIn`

    `signOut`

    `custom`

> -   [Adobe Experience Platform](https://business.adobe.com/products/experience-platform/adobe-experience-platform.html) (requires a subscription and additional merchant [setup](https://experienceleague.adobe.com/docs/experience-platform/edge/fundamentals/datastreams.html?lang=en); data can be used by merchants inside the Adobe Experience Platform for detailed analytics, targeted merchandising, real time customer data profiles, and more)

Reviewer: ```suggestion
-   [Adobe Experience Platform (AEP)](https://business.adobe.com/products/experience-platform/adobe-experience-platform.html) (requires a subscription and additional merchant [setup](https://experienceleague.adobe.com/docs/experience-platform/edge/fundamentals/datastreams.html?lang=en); data can be used by merchants inside the Adobe Experience Platform for detailed analytics, targeted merchandising, real time customer data profiles, and more)
```

    `EventForwarding.aep: true`

    Events not forwarded to AEP Edge:

    `recsItemAddToCartClick`

    `recsItemClick`

    `recsRequestSent`

    `searchResponseReceived`

    `searchCategoryClick`

    `searchProductClick`

    `searchResultsView`

    `searchSuggestionClick`

### Turning off collection per edge

> Events are sent to Adobe Commerce Data Services by default. To turn collection off collection, set `EventForwarding.commerce: false`.

Reviewer: ```suggestion
Events are sent to Adobe Commerce Data Services by default. To turn off event collection, set `EventForwarding.commerce: false`.
```

> Events are not set to AEP Edge unless explicitly configured (`EventForwarding.aep: true`);

Reviewer: ```suggestion
Events are not sent to AEP Edge explicitly. To send events to AEP Edge, set (`EventForwarding.aep: true`);
```

### Handling custom events

> Custom events are supported for the Experience Platform only. Custom data will not be forwarded to Commerce dashboards and metrics trackers.

Reviewer: ```suggestion
Custom events are supported for the Adobe Experience Platform only. Custom data will not be forwarded to Adobe Commerce dashboards and metrics trackers.
```

> For any `custom` event, the collector adds a `personId` (`ecid`) to `customContext` and wraps an `xdm` object around it before forwarding to Egde.

Reviewer: ```suggestion
For any `custom` event, the collector adds a `personId` (`ecid`) to `customContext` and wraps an `xdm` object around it before forwarding to the Edge.
```

Example:

Custom event published through MSE SDK:

```javascript
mse.publish.custom({
    customContext: { customStrAttr: "cheetah", customNumAttr: 128 },
});
```

In AEP Edge:

```javascript
{
    xdm: {
        personId: 'ecid1234',
        customStrAttr: 'cheetah',
        customNumAttr: 128
    }
}
```

> **note** Using custom events may affect OOTB Adobe Analytics reports.

Reviewer: ```suggestion
**Note** Using custom events may affect OOTB Adobe Analytics reports.
```

### Handling event overrides (custom attributes)

> Attribute overrides for standard events are supported for the Experience Platform only. Custom data will not be forwarded to Commerce dashboards and metrics trackers.

Reviewer: ```suggestion
Attribute overrides for standard events are supported for the Adobe Experience Platform only. Custom data will not be forwarded to Adobe Commerce dashboards and metrics trackers.
```

> For any event with a set `customContext`, the collector overrides `personId` and Adobe Analytics counters, and fowards all other attributes set in `customContext`.

Reviewer: ```suggestion
For any event with a set `customContext`, the collector overrides `personId` and Adobe Analytics counters, and forwards all other attributes set in `customContext`.
```

Examples:

Product view with overrides published though MSE SDK:

```javascript
mse.publish.productPageView({
    customContext: { customCode: "okapi" },
});
```

In AEP Edge:

```javascript
{
    xdm: {
        eventType: 'commerce.productViews',
        personId: 'ecid1234',
        customCode: 'okapi',
        commerce: {
            productViews: {
                value : 1
            }
        }
    }
}
```

> Product view with commerce overrides published though MSE SDK:

Reviewer: ```suggestion
Product view with Adobe Commerce overrides published though MSE SDK:
```

```javascript
mse.publish.productPageView({
    customContext: { commerce: { customCode: "mongoose" } },
});
```

In AEP Edge:

```javascript
{
    xdm: {
        eventType: 'commerce.productViews',
        personId: 'ecid1234',
        commerce: {
            customCode: 'mongoose',
            productViews: {
                value : 1
            }
        }
    }
}
```

> **note** Overriding events with custom attributes may affect OOTB Adobe Analytics reports.

Reviewer: ```suggestion
**Note** Overriding events with custom attributes may affect OOTB Adobe Analytics reports.
```
